File: 2019-07-25-Iris-Web-App/server/julia/rest.jl
using HTTP
using JSON

# @info "Loading the model"
include("helpers.jl")

function classify(req::HTTP.Request)    
    @info "Parsing request data"
    data = JSON.parse(String(req.body))
    
    @info "Making prediction"
    _, yidx = findmax(model(data["data"]));
    yprd = yidx == 1 ? "💐 Setosa" : yidx == 2 ? "🌷 Versicolor" : "🌺 Virginica";
    
    @info "Sending response"
    return HTTP.Response(200, [HTTP.setheader(req, "Access-Control-Allow-Origin" => "*")], body = JSON.json(Dict("data" => yprd)), request = req)
end

function train(req::HTTP.Request)
    @info "Requesting to train the model"
    include("./server/julia/model-training.jl")
    accy = accuracy(model, dtst)

    @info "Sending response"
    return HTTP.Response(200, [HTTP.setheader(req, "Access-Control-Allow-Origin" => "*")], body = JSON.json(Dict("data" => accy)), request = req)
end

classifier = HTTP.Router()
HTTP.@register(classifier, "POST", "/classify", classify)

trainer = HTTP.Router()
HTTP.@register(trainer, "POST", "/train", train)

@info "One-time loading of data"
include("load-data.jl")

@info "Listening at 127.0.0.1:8081"
HTTP.listen("127.0.0.1", 8081) do io::HTTP.Stream
    if io.message.target === "/classify"
        HTTP.handle(classifier, io)
    elseif io.message.target === "/train"
        HTTP.handle(trainer, io)
    else
        @info "Oops"
    end
end

File: 2019-07-25-Iris-Web-App/model-deployment/server/python/rest.py
from flask import Flask
from flask import request
from flask import jsonify
import json
import numpy as np

app = Flask(__name__)

@app.route("/classify", methods = ["POST"])
def classify():
    print("Info: Parsing request data")
    data = json.loads(request.get_data().decode())

    print("Info: Making prediction")
    resp = jsonify(classify_helper(data["data"]))

    print("Info: Sending response")
    resp.headers.add("Access-Control-Allow-Origin", "*")
    
    return resp

@app.route("/train", methods = ["POST"])
def train():
    print("Info: Requesting to train the model")
    resp = jsonify(train_helper())

    print("Info: Sending response")
    resp.headers.add("Access-Control-Allow-Origin", "*")

    return resp

def train_helper():
    train_model(xtrn, ytrn, xtst, ytst, batch_size);
    
    return json.dumps({"data" : str(model.evaluate(xtst, ytst)[1])})

def classify_helper(data):
    yidx = np.argmax(model.predict(np.array(data).reshape(1, 4)), axis = 1)[0]
    yprd = "💐 Setosa" if yidx == 0 else "🌷 Versicolor" if yidx == 1 else "🌺 Virginica"

    return json.dumps({"data" : yprd})

if __name__ == "__main__":
    from model_training import model, train_model
    
    print("Info: One-time loading of data")
    from load_data import xtrn, ytrn, xtst, ytst, batch_size

    app.run("127.0.0.1", "8082")